File: src/Loki/Eval.hs
module Loki.Eval
  ( Env (..),
    Value (..),
    CoValue (..),
    eval,
  )
where

import Control.Monad.IO.Class (liftIO)
import Data.Map qualified as M
import Effectful (Eff, IOE, runEff, (:>))
import Effectful.Reader.Static (Reader, ask, local, runReader)
import Loki.Core
import Loki.Prelude

eval :: Env -> Statement Name -> IO ()
eval env stmt = runEff $ runReader env $ evalStatement stmt

evalStatement :: (Reader Env :> es, IOE :> es) => Statement Name -> Eff es ()
evalStatement (Prim name args cont) = do
  args' <- traverse evalProducer args
  cont' <- evalConsumer cont

  case name of
    "add" -> case args' of
      [VInt x, VInt y] -> do
        apply cont' (VInt (x + y))
      _ -> error "invalid arguments"
    _ -> error "unknown primitive"
evalStatement (Switch prod clauses def) = do
  value <- evalProducer prod
  go value clauses
  where
    go _ [] = evalStatement def
    go (VInt m) ((Int n, stmt) : _) | m == n = evalStatement stmt
    go value (_ : rest) = go value rest
evalStatement (Cut (Do label stmt) cont) = do
  covalue <- evalConsumer cont
  local (insertCoValue label covalue) do
    evalStatement stmt
evalStatement (Cut prod cont) = do
  value <- evalProducer prod
  covalue <- evalConsumer cont
  apply covalue value

evalProducer :: (Reader Env :> es) => Producer Name -> Eff es Value
evalProducer (Var name) = do
  env <- ask
  case lookupValue name env of
    Just value -> return value
    Nothing -> error "undefined variable"
evalProducer (Const (Int n)) = return (VInt n)
evalProducer (Do _ _) = error "unexpected statement"

evalConsumer :: (Reader Env :> es) => Consumer Name -> Eff es CoValue
evalConsumer Finish = return CFinish
evalConsumer (Label name) = do
  env <- ask
  case lookupCoValue name env of
    Just value -> return value
    Nothing -> error "undefined label"

apply :: (IOE :> es) => CoValue -> Value -> Eff es ()
apply CFinish x = liftIO $ print x

data Env = Env
  { values :: Map Name Value,
    covalues :: Map Name CoValue
  }
  deriving stock (Show)

instance Semigroup Env where
  Env v1 c1 <> Env v2 c2 = Env (v1 <> v2) (c1 <> c2)

instance Monoid Env where
  mempty = Env mempty mempty

lookupValue :: Name -> Env -> Maybe Value
lookupValue name = M.lookup name . (.values)

lookupCoValue :: Name -> Env -> Maybe CoValue
lookupCoValue name = M.lookup name . (.covalues)

insertCoValue :: Name -> CoValue -> Env -> Env
insertCoValue name value env = env {covalues = M.insert name value env.covalues}

data Value where
  VInt :: Integer -> Value

deriving stock instance Show Value

data CoValue where
  CFinish :: CoValue

deriving stock instance Show CoValue

File: Eval.fs
namespace Loki

open Name

// Eval is a class that evaluates a Core program.
module Eval =
    exception UnknownPrimitive of Location * string
    exception UnexpectedProducer of Location * Core.Producer
    exception UnexpectedConsumer of Location * Core.Consumer
    exception UnknownVariable of Location * Name

    type Value = Int of int

    type Covalue =
        | Finish
        | Mu' of Name * Core.Statement

    type Env =
        { Values: Map<Name, Value>
          Covalues: Map<Name, Covalue> }

        static member empty =
            { Values = Map.empty
              Covalues = Map.empty }

        member this.Lookup loc name =
            match Map.tryFind name this.Values with
            | Some value -> value
            | None -> raise (UnknownVariable(loc, name))

        member this.Add name value =
            { this with
                Values = Map.add name value this.Values }

        member this.LookupCo loc name =
            match Map.tryFind name this.Covalues with
            | Some value -> value
            | None -> raise (UnknownVariable(loc, name))

        member this.AddCo name value =
            { this with
                Covalues = Map.add name value this.Covalues }

    let rec private evalStmt (env: Env) =
        function
        | Core.Prim(loc, name, args, cont) ->
            let args = List.map (evalProducer env) args
            let cont = evalConsumer env cont

            match (name, args) with
            | ("add", [ Int x; Int y ]) -> apply env cont (Int(x + y))
            | ("print", [ Int x ]) ->
                printfn "%d" x
                Int x
            | _ -> raise (UnknownPrimitive(loc, name))
        | Core.Switch(_, prod, clauses, def) ->
            let value = evalProducer env prod

            let rec go clauses =
                match (value, clauses) with
                | (_, []) -> evalStmt env def
                | (Int m, ((Core.Int n, stmt) :: _)) when m = n -> evalStmt env stmt
                | (_, _ :: rest) -> go rest

            go clauses
        | Core.Cut(_, (Core.Mu(_, label, stmt)), cont) ->
            let covalue = evalConsumer env cont
            let env = env.AddCo label covalue
            evalStmt env stmt
        | Core.Cut(loc, prod, cont) ->
            let value = evalProducer env prod
            let covalue = evalConsumer env cont
            apply env covalue value

    and private evalProducer env =
        function
        | Core.Var(loc, name) -> env.Lookup loc name
        | Core.Const(_, Core.Int n) -> Value.Int n
        | p -> raise (UnexpectedProducer(p.Location(), p))

    and private evalConsumer env =
        function
        | Core.Finish _ -> Finish
        | Core.Label(loc, name) -> env.LookupCo loc name
        | Core.Mu'(_, name, stmt) -> Mu'(name, stmt)

    // Apply a value to a continuation.
    and private apply (env: Env) (cont: Covalue) (value: Value) =
        match cont with
        | Finish -> value
        | Mu'(name, stmt) ->
            let env = env.Add name value
            evalStmt env stmt

    let Run (stmt: Core.Statement) = evalStmt Env.empty stmt
